www/_blog/2022-08-11-authentication-tutorial.mdx

```diff
@@ -0,0 +1,597 @@
+
+We will add our two new routes to the router (and remove the default "hello world" one) to get:
+
+```rust
+let router = Router::new()
+    .route("/", get(index))
+    .route("/styles.css", get(styles))
+    .layer(Extension(Arc::new(tera)));
+```
+
+With our main service we can now test it locally with `cargo shuttle run` (go get `cargo install cargo-shuttle`).
+
+![](/images/blog/authentication-demo-screenshot.png)
+
+Nice! 
+
+## Adding users
+
+We will start with a user's table in SQL. ([this is defined in schema.sql](https://github.com/kaleidawave/axum-shuttle-postgres-authentication-demo/blob/main/schema.sql)).
+
+```sql
+CREATE TABLE users (
+    id integer PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
+    username text NOT NULL UNIQUE,
+    password text NOT NULL
+);
+```
+
+The `id` is generated by the database using a sequence. The `id` is a primary key, which we will use to reference users. It is better to use a fixed value field for identification rather than using something like the `username` field because you may add the ability to change usernames, which can leave things pointing to the wrong places.
+
+### Registering our database
+
+Before our app can use the database we have to add sqlx with some features: `cargo add sqlx -F postgres runtime-tokio-native-tls`. We will also enable the Postgres feature for shuttle with `cargo add shuttle-service -F sqlx-postgres`. 
+
+Now back in the code we use `#[shared::Postgres]`. This annotation tells shuttle to provision a Postgres database using the [infrastructure from code design](https://www.shuttle.rs/blog/2022/05/09/ifc)!
+
+```rust
+type Database = sqlx::PgPool;
+
+#[shuttle_service::main]
+async fn axum(
+    #[shared::Postgres] pool: Database
+) -> ShuttleAxum {
+    let router = Router::new()
+        .route("/", get(index))
+        .route("/styles.css", get(styles))
+        .layer(Extension(Arc::new(tera)))
+        .layer(database)
+    Ok(sync_wrapper::SyncWrapper::new(get_router(pool)))
+}
+```
+
+### Signup
+
+For getting users into our database, we will create a post handler. In our handler, we will parse data using multipart. [I wrote a simple parser for multipart that we will use here](https://github.com/kaleidawave/axum-shuttle-postgres-authentication-demo/blob/main/src/utils.rs#L45-L64). The below example contains some error handling that we will ignore for now.
+
+```rust
+async fn post_signup(
+    Extension(database): Extension<Database>,
+    multipart: Multipart,
+) -> impl IntoResponse {
+    let data = parse_multipart(multipart)
+        .await
+        .map_err(|err| error_page(&err))?;
+
+    if let (Some(username), Some(password), Some(confirm_password)) = (
+        data.get("username"),
+        data.get("password"),
+        data.get("confirm_password"),
+    ) {
+        if password != confirm_password {
+            return Err(error_page(&SignupError::PasswordsDoNotMatch));
+        }
+
+        // We will create users here
+
+        Ok(())

```

Comment:
It's a little hard to follow how the code pieces below need to be added here. Maybe this should be a function call (`try_to_add_user()`) and the code pieces below be in the block of this called function?

Reply:
Or maybe there should be three comment lines here?
1. Hash password
1. Insert user and get user id
1. Create session and set cookie

Reply:
Or 3+ function calls with the code pieces below defining each function :man_shrugging:

Reply:
The function calls suggestion also ties in with the `set_cookie` refactor comment